Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StructRendering.kt

```diff
@@ -0,0 +1,94 @@
+package org.jetbrains.kotlin.native.interop.gen
+
+import org.jetbrains.kotlin.native.interop.indexer.*
+
+fun tryRenderStructOrUnion(def: StructDef): String? = when (def.kind) {
+    StructDef.Kind.STRUCT -> tryRenderStruct(def)
+    StructDef.Kind.UNION -> tryRenderUnion(def)
+}
+
+private fun tryRenderStruct(def: StructDef): String? {
+    val isPackedStruct = def.fields.any { !it.isAligned }
+
+    var offset = 0L
+
+    return buildString {
+        append("struct")
+        if (isPackedStruct) append(" __attribute__((packed))")
+        append(" { ")
+
+        def.members.forEachIndexed { index, it ->
+            val name = "p$index"
+            val decl = when (it) {
+                is Field -> {
+                    if (it.offset % 8L == 0L) {
+
+                        val defaultAlignment = if (isPackedStruct) 1L else it.typeAlign
+                        val alignment = guessAlignment(offset, it.offsetBytes, defaultAlignment) ?: return null
+
+                        offset = it.offsetBytes + it.typeSize
+
+                        tryRenderVar(it.type, name)
+                                ?.plus(if (alignment == defaultAlignment) "" else "__attribute__((aligned($alignment)))")
+                    } else {
+                        null // e.g. flexible array member. TODO.
+                    }
+                }
+                is BitField -> null // TODO: tryRenderVar(it.type, name)?.plus(" : ${it.size}")
+            } ?: return null
+            append("$decl; ")
+        }
+
+        append("}")
+    }
+}
+
+private fun guessAlignment(offset: Long, paddedOffset: Long, defaultAlignment: Long): Long? =
+        longArrayOf(defaultAlignment, 1L, 2L, 4L, 8L, 16L, 32L).firstOrNull {
+            alignUp(offset, it) == paddedOffset
+        }
+
+private fun alignUp(x: Long, alignment: Long): Long = (x + alignment - 1) and (alignment - 1).inv()
+
+private fun tryRenderUnion(def: StructDef): String? =
+        if (
+                def.members.any {
+                    when (it) {
+                        is Field -> it.offset != 0L
+                        is BitField -> it.offset != 0L
+                    }
+                }
+        ) {
```

Comment:
Maybe symmetry of `{}` in then/else blocks?

Reply:
Done.